content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+
+   - `Relay` can limit the resources used to relay connections (e.g., by the number 
+     of connections, the time, and bytes) via Circuit Relay v2. In the case of IPFS, 
+     this allows every public node in the network to serve as a relay without high 
+     resource consumption.
+   - For each discovered `Relay`, `B`: 
+       - connects to the remote node and requests the Relay node to listen to 
+         connections on its behalf, known as a reservation;
+       - if `Relay` accepts reservation requests, `B` can advertise itself as being 
+         reachable through `Relay`.
+
+### Phase II: Hole punching
+
+1. [Circuit Relay](/concepts/circuit-relay): Establish a secure relay connection 
+   through the public Relay node. The node establishes a direct connection with 
+   the Relay node. It then requests a relayed connection to the other node through 
+   the Relay node, creating a bi-directional channel and using TLS to secure the 
+   channel.
+   
+    ![](https://i.imgur.com/zoIWcwK.png)
+
+    - `A` establishes a relayed connection to `B` via the `Relay` using the 
+      information contained in `B`'s advertised address. 
+        - `A` first establishes a direct connection to `Relay` and then 
+          requests a relayed connection to `B` from `Relay`. 
+        - `Relay` forwards said request to `B` and accepts. 
+        - `Relay` forwards the acceptance to `A`. 
+        - `A` and `B` can use the bi-directional channel over `Relay` to 
+          communicate.
+        - `A` and `B` upgrade the relayed connection with a security protocol 
+          like TLS.
+   <!-- to add dcutr reference when available -->
+2. [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md): Use 
+   DCUtR as a synchronization mechanism to coordinate hole punching.
+
+    ![](https://i.imgur.com/wXehUlC.png)
+
+    - `A` sends a `Connect` message to `B` through `Relay`. 
+        - `Connect` contains the addresses of A. libp2p and offers multiple 
+          mechanisms to discover one's addresses, e.g., via the libp2p identify 
+          protocol. 
+    - `B` receives the `Connect` message on the relayed connection and replies 
+      with a `Connect` message containing its (non-relayed) addresses. 
+    - `A` measures the time between sending its message and receiving `B`'s 
+      message, thereby determining the round-trip time between `A` and `B` via `Relay`.
+    - Then, `A` sends a `Sync` message to `B` on the relayed connection. 
+    - `A` waits for half the round-trip time, then directly dials `B` via the 
+      addresses received in `B`'s `Connect`.
+    - As soon as `B` receives `A`'s `Sync` message, it directly dials `A` with the 
+      addresses provided in `A`'s `Connect` message.
+    - Once `A` and `B` dial each other simultaneously, a hole punch occurs.
+
+### Resources
+
+- This guide is a byproduct of the 
+  [Hole punching in libp2p - Overcoming Firewalls](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/)
```

Comment:
We should link the DINPS paper here. I’ll get you the link.

Reply:
@DannyS03 I think this is the paper https://research.protocol.ai/publications/decentralized-hole-punching/